### PR TITLE
logging: implement lazy sprinting

### PR DESCRIPTION
**Pull request by tychoish**
shout out to @joeabbey for the inspiration. This makes the lazy
functions internal by default to prevent potential misuse by external
callers.

Should backport cleanly into 0.36 and I'll handle a messy merge into 0.35

### CODE REVIEW

**Review by joeabbey — COMMENTED** [review]
Can this be internal and unexported?

Rephrased, how will cosmos-sdk satisfy the LogObjectMarshaler interface for [zerolog](https://github.com/rs/zerolog)?
```
func (l *LazySprintf) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.String())
}

func (l *LazyBlockHash) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.Block.Hash().String())
}
```

**Comment by tychoish** [comment]
> Can this be internal and unexported?

I believe that it is: the new lazy functions are in an internal function and not accessible outside of Tendermint, which is what I think we decided as a group. 

> Rephrased, how will cosmos-sdk satisfy the LogObjectMarshaler interface for [zerolog](https://github.com/rs/zerolog)?

Tendermint's logging library just passes types through to zerolog. I don't think this changes any thing, so the answer is "the same way it's always been done."

**Comment by tychoish** [comment]
I've been mucking around a bit in zerologger and our logging implementation, I think that this won't actually save much time, as the way zerolog is going to handle a value that's a fmt.Sprintf() is the same as what you'd get by adding these methods to the type, but in a way that is a bit awkward with the extra field and nesting. 

```go
func (l *LazySprintf) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.String())
}

func (l *LazyBlockHash) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.Block.Hash().String())
}
```

Related to this change, we were previously iterating through all of the fields for all log messages to form a map that we would hand to zerolog, when zerolog could handle our field formats just fine on its own, and I did that in a separate PR that's already merged, so while I don't think that will save us *that* much time (we have a handful of fields in any given log site,) it'll probably help eventually as well.

File: internal/libs/strings/string.go [hunk 1]
```diff
@@ -3,8 +3,64 @@ package strings
...
+func (s *lazyStringf) String() string {
+	if s.out == "" && s.tmpl != "" {
+		s.out = fmt.Sprintf(s.tmpl, s.args)
+		s.args = nil
+		s.tmpl = ""
+	}
+	return s.out
+}
+
+func LazySprintf(t string, args ...interface{}) fmt.Stringer {
```
Comment (thanethomson) [inline internal/libs/strings/string.go:25]: I think we need some documentation for this, including a big warning reminding people that the result captures the arguments and is not thread-safe.

File: internal/libs/strings/string.go [hunk 2]
```diff
@@ -3,8 +3,64 @@ package strings
...
+
+func (l *lazyStringer) String() string {
+	if l.out == "" && l.val != nil {
+		l.out = l.val.String()
+		l.val = nil
+	}
+	return l.out
+}
+
+func LazyStringer(v fmt.Stringer) fmt.Stringer { return &lazyStringer{val: v} }
```
Comment (thanethomson) [inline internal/libs/strings/string.go:42]: Same as above regarding documenting this.

File: internal/libs/strings/string.go [hunk 3]
```diff
@@ -3,8 +3,64 @@ package strings
...
+
+type lazyBlockHash struct {
+	block interface{ Hash() tmbytes.HexBytes }
+	out   string
+}
+
+// NewLazyBlockHash defers block Hash until the Stringer interface is invoked.
+// This is particularly useful for avoiding calling Sprintf when debugging is not
+// active.
```
Comment (thanethomson) [inline internal/libs/strings/string.go:51]: I think it'd also be useful to note here that the result captures the argument and is not thread-safe.

**Comment by tychoish** [comment]
@thanethomson so I think that the discussions we've had about this might have actually over-complicated this. These functions are thread safe, and the evaluation semantics of arguments are *the same as every other go function*, the difference is that the string is resolved when (and if) the logging statement is executed and not when the statement is evaluated. The type is just as thread safe as any other type in Go, and the semantics of these types are the same as any other function call in go, and in some ways documenting this could be more confusing rather than less confusing. 

I've added some additional documentation.

**Review by creachadair — APPROVED** [review]

**Pull request merged**